### DISCUSSION TITLE
Need help getting simple context-dependent instruction working

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#7002] [Q&A]
I'm implementing a processor for a 40-year old bytecode interpreter (GPL for the TI-99/4A) which has two modes and I'm struggling to implement the modal part. It's much simpler than the example in the Sleigh documentation and I'm unclear as to what to leave out to adapt to my case. Here's the relevant parts of my `.slaspec`

```
define register offset=0x40 size=4 contextreg;

define context contextreg
    in_fmt = (0,1)
;

# enter format mode
:FMT is op=0x08 [in_fmt=1] { }

# exit format mode
:FEND in_fmt is in_fmt=1 & op=0xFB [in_fmt=0;] { }

# the first formatting instruction that occurs after a `FMT` in my sample code
:ROW p1s is in_fmt=1 & op=0xFE ;p1s { }
```

Of course the `FMT` instruction appears in the disassembly but the `FEND` and `ROW` instructions etc are never recognized.

<img width="701" alt="image" src="https://github.com/user-attachments/assets/99b73d9b-f139-4224-a2e5-041d7d28fbf8">

If I add test instructions with the same opcode but don't check `in_fmt` then that gets recognized instead, which makes it seem that `in_fmt` is not getting set to `1` as intended, or it's getting reset back to `0` somehow. I don't know how to check/debug that case.

I tried making `in_fmt` bigger and incrementing and decrementing it inside `FMT` and `FEND` but that made no difference. (When I make it 32-bit the loader gets stuck in an infinite loop.)

What am I missing / doing wrong here?

**Comment by GhidorahRex** [comment]
I'll have to look closer at the sleigh code you've provided to solve the issue, but we *do* have a script that could help in Ghidra. Put the cursor on the first byte of the instruction you're interested in debugging and open the script menu. The script you want to run is `DebugSleighInstructionParse` - I use it enough that I have mapped it to a keyboard shortcut. It can be a little cumbersome to work through the output, but it is helpful for navigating instruction errors (either not parsing, or parsing to the wrong instruction).

**Reply by hippietrail** [reply]
Aha! I've actually never experimented with the scripts yet. Maybe because I dove straight into the Java stuff and then the Sleigh stuff I didn't pay any attention to any of the Python stuff, which I assumed all the scripts were relating to.

**Comment by hippietrail** [comment]
It turns out there was more to how modal/context works, and I have it working now by adding `globalset(inst_next,in_fmt);` after modifying `in_fmt`. This makes it work globally, which I what I was expecting. (I don't understand the local case for context yet.)

```
:FMT  is op=0x08            [in_fmt=in_fmt+1; globalset(inst_next,in_fmt);] { }
:FEND is in_fmt>0 & op=0xFB [in_fmt=in_fmt-1; globalset(inst_next,in_fmt);] { }
```
